crates/adapters/src/server/mod.rs

```diff
@@ -178,47 +200,62 @@ impl ErrorResponse {
     {
         let message = error.to_string();
         let error_code = error.error_code();
+        let details = serde_json::to_value(error).unwrap_or_else(|e| {
+            JsonValue::String(format!("Failed to serialize error. Details: '{e}'"))
+        });
 
         error!("[HTTP error response] {error_code}: {message}");
+        // Uncomment this when all pipeline manager errors implement `ResponseError`
+        // if error.status_code() == StatusCode::INTERNAL_SERVER_ERROR {
+        if let Some(backtrace) = details.get("backtrace").and_then(JsonValue::as_str) {
+            error!("Error backtrace:\n{backtrace}");
+        }
+        // }
 
         Self {
             message,
             error_code,
-            details: serde_json::to_value(error).unwrap_or_else(|e| {
-                JsonValue::String(format!("Failed to serialize error. Details: '{e}'"))
-            }),
+            details,
         }
     }
-}
 
-fn http_resp_from_controller_error(error: &ControllerError) -> HttpResponse {
-    match error {
-        ControllerError::Config {
-            config_error: ConfigError::UnknownInputStream { .. },
-        } => HttpResponse::NotFound(),
-        ControllerError::Config {
-            config_error: ConfigError::UnknownOutputStream { .. },
-        } => HttpResponse::NotFound(),
-        ControllerError::Config { .. } => HttpResponse::BadRequest(),
-        ControllerError::ParseError { .. } => HttpResponse::BadRequest(),
-        ControllerError::PipelineTerminating { .. } => HttpResponse::NotFound(),
-        _ => HttpResponse::InternalServerError(),
+    pub fn from_error_nolog<E>(error: &E) -> Self
```

Comment:
Rather than duplicate from_error() in from_error_nolog(), from_error() can just do something like `let err = from_error_nolog(); error!("....{}; {}", err.error_code, err.message)` instead.